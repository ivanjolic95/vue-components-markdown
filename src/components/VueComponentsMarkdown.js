import VueMarkdown from 'vue-markdown'

function getPropValue(prop, variables) {
  const propValue = prop.split(': ')[1].trim().replace(/'/g, '')
  const isVariable = propValue.match(/^{{\w*}}$/)
  if (isVariable) {
    return variables[propValue.replace(/{|}/g, '')]
  }
  return JSON.parse(propValue)
}

function extractComponentConfig(component, variables) {
  // split only by first space
  const componentConfig = component
    .match(/[^!<>]*/g)
    .filter(Boolean)[0]
    .match(/^(\S+)\s(.*)/)
    .slice(1)
  const componentName = componentConfig[0]
  const componentProps = componentConfig[1]
    .slice(1, -1)
    .split(', ')
    .map((prop) => prop.trim())
    .reduce((val, item) => {
      const propName = item.split(':')[0]
      const propValue = getPropValue(item, variables)
      return { ...val, [propName]: propValue }
    }, {})
  return { componentName, componentProps }
}

function renderComponent(
  h,
  component,
  componentMap,
  variables,
  noComponentWrapper,
  componentWrapperTag,
  componentWrapperClass
) {
  const { componentName, componentProps } = extractComponentConfig(
    component,
    variables
  )
  const componentElement = h(
    componentMap[componentName],
    { props: componentProps },
    []
  )
  return noComponentWrapper
    ? componentElement
    : h(componentWrapperTag, { staticClass: componentWrapperClass }, [
        componentElement,
      ])
}

function splitMarkdownContentAndComponents(content) {
  const COMPONENT_REGEX = /(!<.*>)/g

  return (
    content.split(COMPONENT_REGEX).map((paragraph) => ({
      isComponent: !!paragraph.match(COMPONENT_REGEX),
      content: paragraph,
    })) || []
  )
}

function injectComponents(
  h,
  {
    content,
    componentMap,
    variables,
    noComponentWrapper,
    componentWrapperTag,
    componentWrapperClass,
  },
  scopedSlots
) {
  const paragraphs = splitMarkdownContentAndComponents(content)

  const elements = paragraphs.map((paragraph) =>
    paragraph.isComponent
      ? renderComponent(
          h,
          paragraph.content,
          componentMap,
          variables,
          noComponentWrapper,
          componentWrapperTag,
          componentWrapperClass
        )
      : (scopedSlots['markdown-processor'] &&
          scopedSlots['markdown-processor']({ content: paragraph.content })) ||
        h(VueMarkdown, { props: { source: paragraph.content } }, [])
  )

  return h('div', { staticClass: 'VueComponentsMarkdown' }, elements)
}

function renderBlocks(h, props, scopedSlots) {
  return injectComponents(h, props, scopedSlots)
}

export default {
  name: 'VueComponentsMarkdown',
  functional: true,
  props: {
    content: {
      type: String,
      default: '',
    },
    componentMap: {
      type: Object,
      default: () => ({}),
    },
    variables: {
      type: Object,
      default: () => ({}),
    },
    noComponentWrapper: {
      type: Boolean,
      default: false,
    },
    componentWrapperTag: {
      type: String,
      default: 'div',
    },
    componentWrapperClass: {
      type: String,
      default: '',
    },
  },
  render(h, { props, scopedSlots }) {
    return renderBlocks(h, props, scopedSlots)
  },
}
