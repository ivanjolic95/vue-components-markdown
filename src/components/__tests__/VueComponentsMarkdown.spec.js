import Vue from 'vue'
import { mount, shallowMount } from '@vue/test-utils'

import VueMarkdown from 'vue-markdown'

import VueComponentsMarkdown from '../VueComponentsMarkdown'

const MyComponent = Vue.extend({
  name: 'MyComponent',
  props: {
    title: {
      type: String,
      required: true,
    },
    text: {
      type: String,
      required: true,
    },
  },
  render(h) {
    return h('div', { staticClass: 'MyComponent' }, [
      h('h1', [this.title]),
      h('p', [this.text]),
      h('button', ['My button']),
    ])
  },
})

describe('VueComponentsMarkdown', () => {
  let wrapper
  let content
  let variables

  beforeEach(() => {
    content = `
# This is a title

This is a __test__.

!<my-component { title: 'Title', text: '{{customText}}' }>

Another regular markdown paragraph. This time with list items:
- first
- second
- third
  - subitem

And back to the custom component:
!<my-component { title: 'Title', text: '{{customText}}' }>`

    variables = {
      customText: 'This is my custom text passed as a variable.',
    }

    wrapper = shallowMount(VueComponentsMarkdown, {
      propsData: {
        content,
        componentMap: {
          'my-component': MyComponent,
        },
        variables,
      },
    })
  })

  describe('Rendering', () => {
    beforeEach(() => {
      wrapper = mount(VueComponentsMarkdown, {
        propsData: {
          content,
          componentMap: {
            'my-component': MyComponent,
          },
          variables,
        },
      })
    })
    test('renders properly', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('Injecting custom components', () => {
    test('shows the custom component', () => {
      expect(wrapper.findComponent(MyComponent).exists()).toBeTruthy()
    })

    test('sets prop with static value', () => {
      expect(wrapper.findComponent(MyComponent).props().title).toBe('Title')
    })

    test('sets prop with dynamic value', () => {
      expect(wrapper.findComponent(MyComponent).props().text).toBe(
        'This is my custom text passed as a variable.'
      )
    })

    test('uses `div` as wrapper element for custom components by default', () => {
      expect(
        wrapper.findComponent(MyComponent).element.parentElement.tagName
      ).toBe('DIV')
    })

    test('uses `componentWrapperTag` as wrapper element for custom components', () => {
      wrapper = mount(VueComponentsMarkdown, {
        propsData: {
          content,
          componentMap: {
            'my-component': MyComponent,
          },
          variables,
          componentWrapperTag: 'p',
        },
      })
      expect(
        wrapper.findComponent(MyComponent).element.parentElement.tagName
      ).toBe('P')
    })

    test('uses `componentWrapperClass` as class attribute for wrapper element for custom components', () => {
      wrapper = mount(VueComponentsMarkdown, {
        propsData: {
          content,
          componentMap: {
            'my-component': MyComponent,
          },
          variables,
          componentWrapperClass: 'test-class',
        },
      })
      expect(
        wrapper.findComponent(MyComponent).element.parentElement.classList
      ).toContain('test-class')
    })
  })

  describe('using custom markdown processor', () => {
    test('uses `VueMarkdown` for processing markdown by default', () => {
      expect(wrapper.findComponent(VueMarkdown).exists()).toBeTruthy()
    })

    test('uses custom markdown processor if provided as `markdown-processor` scoped slot', () => {
      wrapper = shallowMount(VueComponentsMarkdown, {
        propsData: {
          content,
          componentMap: {
            'my-component': MyComponent,
          },
          variables,
        },
        scopedSlots: {
          'markdown-processor':
            '<p>This is a custom markdown processor: {{props.content}}</p>',
        },
      })

      expect(wrapper.text()).toContain('This is a custom markdown processor')
    })
  })
})
