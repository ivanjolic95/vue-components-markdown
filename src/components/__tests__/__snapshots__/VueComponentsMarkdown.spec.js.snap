// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueComponentsMarkdown Rendering renders properly 1`] = `
<div
  class="VueComponentsMarkdown"
>
  <div>
    <h1
      id="this-is-a-title"
      tabindex="-1"
    >
      <a
        class="markdownIt-Anchor"
        href="#this-is-a-title"
      >
        #
      </a>
       This is a title
    </h1>
    

    <p>
      This is a 
      <strong>
        test
      </strong>
      .
    </p>
    

  </div>
  <div
    class=""
  >
    <div
      staticclass="MyComponent"
    >
      <h1>
        Title
      </h1>
      <p>
        This is my custom text passed as a variable.
      </p>
      <button>
        My button
      </button>
    </div>
  </div>
  <div>
    <p>
      Another regular markdown paragraph. This time with list items:
    </p>
    

    <ul>
      

      <li>
        first
      </li>
      

      <li>
        second
      </li>
      

      <li>
        third

        <ul>
          

          <li>
            subitem
          </li>
          

        </ul>
        

      </li>
      

    </ul>
    

    <p>
      And back to the custom component:
    </p>
    

  </div>
  <div
    class=""
  >
    <div
      staticclass="MyComponent"
    >
      <h1>
        Title
      </h1>
      <p>
        This is my custom text passed as a variable.
      </p>
      <button>
        My button
      </button>
    </div>
  </div>
  <div />
</div>
`;
